torchaudio/datasets/cmudict.py

```diff
@@ -0,0 +1,105 @@
+    """
+
+    def __init__(self,
+                 root: Union[str, Path],
+                 url: str = "http://svn.code.sf.net/p/cmusphinx/code/trunk/cmudict/cmudict-0.7b",
+                 url_symbols: str = "http://svn.code.sf.net/p/cmusphinx/code/trunk/cmudict/cmudict-0.7b.symbols",
+                 download: bool = False,
+                 exclude_punctuations: bool = True) -> None:
+
+        self.exclude_punctuations = exclude_punctuations
+
+        root = Path(root)
+        if not os.path.isdir(root):
+            os.mkdir(root)
+
+        if download:
+            if os.path.isdir(root):
+                checksum = _CHECKSUMS.get(url, None)
+                download_url(url, root, hash_value=checksum, hash_type="md5")
+                checksum = _CHECKSUMS.get(url_symbols, None)
+                download_url(url_symbols, root, hash_value=checksum, hash_type="md5")
+            else:
+                RuntimeError(f"The argument `root` must be a path to directory, "
+                             "but '{root}' is passed in instead.")
+
+        self._root_path = root
+        basename = os.path.basename(url)
+        basename_symbols = os.path.basename(url_symbols)
+
+        with open(os.path.join(self._root_path, basename_symbols), "r") as text:
+            self._symbols = text.readlines()
+
+        with open(os.path.join(self._root_path, basename), "r") as text:
+            self._dictionary = self._parse_dictionary(text.readlines())
+
+    def _parse_dictionary(self, lines: List[str]):
+        _alt_re = re.compile(r'\([0-9]+\)')
+        cmudict: List[Tuple[str, List[str]]] = list()
+        for line in lines:
+            if not line or line.startswith(';;;'):  # ignore comments
+                continue
+
+            word, phones = line.strip().split('  ')
+            if word[0] in string.punctuation:
+                if self.exclude_punctuations:
+                    continue
+                else:
+                    # !EXCLAMATION-POINT -> !
+                    word = word[0]
+
+            # if a word have multiple pronunciations, there will be (number) appended to it
+            # for example, DATAPOINTS and DATAPOINTS(1),
+            # the regular expression `_alt_re` removes the '(1)' and change the word DATAPOINTS(1) to DATAPOINTS
+            word = re.sub(_alt_re, '', word)
+            phones = phones.split(" ")
+            cmudict.append((word, phones))
+        return cmudict
+
+    def __getitem__(self, n: int) -> Tuple[str, List[str]]:
+        """Load the n-th sample from the dataset.
+
+        Args:
+            n (int): The index of the sample to be loaded.
+
+        Returns:
+            tuple: The corresponding word and phonemes ``(word, [phonemes])``.
+
+        """
+        return self._dictionary[n]
+
+    def __len__(self) -> int:
+        return len(self._dictionary)
+
+    def get_symbols(self) -> List[str]:
```

Comment:
1. Is `List[str]` the right data structure? I am getting a feeling that `set` is more appropriate, and if that's the case, we can use `frozenset` so that we do not need to copy.
2. Can we use `@property` access? If the underlying symbols are frozen, it is more Pythonic to access it as a property, rather than getter/setter approach.

Reply:
For 1, alternative way to think is; do we need to iterate the symbols?

Reply:
Currently in the dataset, we don't need to iterate the symbols. But in the normal usage, we will need to translate each character into a number, that is where we need to iterate the symbols.

The symbol here is very similar the the symbol [here](https://github.com/pytorch/audio/blob/master/examples/pipeline_tacotron2/text/text_preprocessing.py#L65). And we will need to map each symbol (character) to a number like [here](https://github.com/pytorch/audio/blob/master/examples/pipeline_tacotron2/text/text_preprocessing.py#L85) and [here](https://github.com/pytorch/audio/blob/master/examples/pipeline_tacotron2/text/text_preprocessing.py#L85). So we actually need their order.

Reply:
So in client code, the ordering is used for indexing, but the question is, the ordering in client code has to be natively originated from the `CMUDict` class?

For example, I see a mapping "ABE -> EY1 B" in the dictionary. But the class `CMUDict` does not define/care to which index `EY1` or `B` should be mapped. `CMUDict` only defines G2P mapping and the set of phoneme symbols. It appears to me that how to index the symbols is user's choice. If we use `List[str]`, then we are giving that this is the order that the symbols should be used.

Reply:
No, it does not need to be natively originated from the `CMUDict` class. The users can always turn the set into a list and sort it in the way they like. 

And yes, I agree that using `List[str]` is similar to giving the user instruction on the order of the symbols (another way of looking at it is that we are preserving the order information from the original dataset). But is there a possible downside of having a default ordering of the symbols?